lib/puppetserver/ca/create_action.rb

```diff
@@ -0,0 +1,268 @@
+      # them, download the signed certificates from the CA, and finally save
+      # the signed certs and associated keys. Returns true if all certs were
+      # successfully created and saved.
+      def generate_certs(certnames, settings, digest)
+        passed = certnames.map do |certname|
+          key, csr = generate_key_csr(certname, settings, digest)
+          return false unless submit_certificate_request(certname, csr.to_s, settings)
+          return false unless sign_cert(certname, settings)
+          if download_cert(certname, settings)
+            save_keys(key, certname, settings)
+            true
+          else
+            false
+          end
+        end
+        passed.all?
+      end
+
+      def generate_key_csr(certname, settings, digest)
+        host = Puppetserver::Ca::Host.new(digest)
+        private_key = host.create_private_key(settings[:keylength])
+        csr = host.create_csr(certname, private_key)
+
+        return private_key, csr
+      end
+
+      def http_client(settings)
+        @client ||= RefactoredClient.new(settings[:localcacert],
+                                         settings[:certificate_revocation],
+                                         settings[:hostcrl])
+      end
+
+      # Make an HTTP request to submit certificate requests to CA
+      # @param certname [Array] the name of the certificate to fetch
+      # @param csr [String] string version of a OpenSSL::X509::Request
+      # @param settings [Hash] a hash of config settings
+      # @return [Boolean] success of all csrs being submitted to CA
+      def submit_certificate_request(certname, csr, settings)
+        client = http_client(settings)
+        url = client.make_ca_url(settings[:ca_server],
+                       settings[:ca_port],
+                       'certificate_request')
+
+        client.with_connection(url) do |connection|
+          url.resource_name = certname
+          result = connection.put(csr, url)
+          check_submit_result(result, certname)
+        end
+      end
+
+      def check_submit_result(result, certname)
+        case result.code
+        when '200', '204'
+          @logger.inform "Successfully submitted certificate request for #{certname}"
+          return true
+        when '400'
+          @logger.err 'Error:'
+          @logger.err "    Sorry, looks like a signed certificate already exists for #{certname}"
+          @logger.err "    Can't create a cert for a certname when one already exists."
+          return false
+        else
+          @logger.err 'Error:'
+          @logger.err "    When certificate request submitted for #{certname}:"
+          @logger.err "      code: #{result.code}"
+          @logger.err "      body: #{result.body.to_s}" if result.body
+          return false
+        end
+      end
+
+      # Make an HTTP request to CA to sign the named certificates
+      # @param certname [Array] the name of the certificate to have signed
```

Comment:
This should be a `String` right?

Reply:
Yes, thank you. Left over from when it was an array